File: Baekjoon/2448/2448.py
def init(star_list, n):
        for i in range(n):
                star_list.append(" " * (2 * n -1))
        for i in range(n):
                star_list[i] = star_list[i][:n - i - 1] + "*"*(2 * i + 1) + star_list[i][n + i:]

def erase_pyramid(star_list, row, col, height):
        if height <= 3:
                star_list[row + 1] = star_list[row + 1][:col] + " " + star_list[row + 1][col + 1:] 
                # print(star_list[row])
        else:
                erase_range = int(height/2 - 1)
                erase_row_s = int(row + (height) / 2)
                erase_col_s = int(col)
                for y in range(erase_row_s, row + height):
                        star_list[y] = star_list[y][:erase_col_s - erase_range] + " "*(2*erase_range + 1) + star_list[y][erase_col_s + erase_range + 1:]
                        erase_range = erase_range - 1
                erase_pyramid(star_list, row, col, int(height/2))
                erase_pyramid(star_list, row + int(height/2), col - int(height/2), int(height/2))
                erase_pyramid(star_list, row + int(height/2), col + int(height/2), int(height/2))

N = int(input())
star_arr = []
init(star_arr, N)
erase_pyramid(star_arr, 0, N - 1, N)
for line in star_arr:
        print(line)

File: Baekjoon/2448/2448.cc
#include <iostream>

void init(char **, int);
void erase_pyramid(char **, int, int, int);
int N = 0;

int main(void) {
    std::cin >> N;
    char ** pyramid = new char*[N];
    for (int i = 0; i < N; i++) {
        pyramid[i] = new char[2* N + 1];
        for (int j = 0; j < 2*N; j++) {
            pyramid[i][j] = ' ';
        }
        pyramid[i][2 * N - 1] = '\0';
    }
    init(pyramid, N);
    erase_pyramid(pyramid, 0, N - 1, N);
    for (int i = 0; i < N; i++)
        std::cout << pyramid[i] << '\n';
}

void init(char ** py_, int n) {
    for (int i = 0; i < n; i++) {
        for (int j = -i; j <= i; j++) {
            py_[i][n - 1 + j] = '*';
        }
    }
}

void erase_pyramid(char ** py_, int row, int col, int height) {
    // std::cout << "row : " << row << " , col is : " << col << " , height is : " << height << '\n';
    // py_[row][col] = '@';
    if (height <= 3) {
        py_[row + 1][col] = ' ';
        // for (int i = 0; i < N; i++)
        //     std::cout << py_[i] << '\n';
    } else {
        int erase_range = ((height) / 2) - 1;
        int erase_row_s = row + (height) / 2;
        int erase_col_s = col;
        for (int y = erase_row_s; y < row + height; y++) {
            for (int x = erase_col_s - erase_range; x <= erase_col_s + erase_range; x++) {
                py_[y][x] = ' ';
            }
            erase_range--;
        }
        // for (int i = 0; i < N; i++)
        //     std::cout << py_[i] << '\n';
        erase_pyramid(py_, row, col, height/2);
        erase_pyramid(py_, row + height/2, col - height/2, height/2);
        erase_pyramid(py_, row + height/2, col + height/2, height/2);
    }
}